### ISSUE TITLE
Passing an iteration object by reference to a function

### ISSUE THREAD

**Issue by J-e-r-e-m-i-a-h** [nlohmann/json#967]
**Bug Report**

- What is the issue you have?
I cannot pass a JSON iteration object by reference to a function

- Please describe the steps to reproduce the issue. Can you provide a small but working code example?
```
nlohmann::json j;
std::ifstream if( "C:\\example.json" );
j << if;
if.close( );

/*...*/

void square( int &in )
{
    in *= in;
}

/*...*/

square( &j[ "int" ] );
```
- What is the expected behavior?
To be able to pass the object by reference, rather than this:
```
int buf = j[ "int" ];
square( &buf );
j[ "int" ] = buf;
```
- And what is the actual behavior instead?
`Error (active)    E0167    argument of type "nlohmann::basic_json<std::map, std::vector, std::string, bool, int64_t, uint64_t, double, std::allocator, nlohmann::adl_serializer> *" is incompatible with parameter of type "int *"`

- Which compiler and operating system are you using? Is it a [supported compiler](https://github.com/nlohmann/json#supported-compilers)?
Visual Studio 2017 15.5.3, Windows 10

- Did you use a released version of the library or the version from the `develop` branch?
Released

- If you experience a compilation error: can you [compile and run the unit tests](https://github.com/nlohmann/json#execute-unit-tests)?
Yes

**Comment by GTXZombie25** [comment]
+1

**Comment by pboettch** [comment]
Your method prototype `void square(int &in)` expects a reference (`&`) and you're calling it by passing a pointer to a json-object (`&buf` gives a `json *` as very simplified explanation).

Try removing the `&` of `buf`:

```CPP
int buf = j[ "int" ];
square(buf);
j["int"] = buf;
```

**Comment by pboettch** [comment]
One problem is that you're mixing references and pointers. Your `square`-function expects a reference to in int (`int &`) but you're passing a pointer (`&j["int"]`). (The same is true for your "rather than this"-example, which won't work either).

**Comment by nlohmann** [comment]
You can use the [`get_ref` function](https://nlohmann.github.io/json/classnlohmann_1_1basic__json_afbd800010b67619463c0fce6e74f7878.html#afbd800010b67619463c0fce6e74f7878) to get a reference and the [`get_ptr` function](https://nlohmann.github.io/json/classnlohmann_1_1basic__json_aefa46bd2d96bb77a38d1c8b431eab44f.html#aefa46bd2d96bb77a38d1c8b431eab44f) to get a point to a stored value.

However, the library does not store integers as `int`, but uses `uint64_t` for unsigned values and `int64_t` for signed values. You can check the value with `is_number_unsigned()` and `is_number_integer()`, respectively. In either case, `get_ref<int&>()` or `get_ptr<int*>()` will not work.

You could template your `square` function to work with any number reference type, and then pass a reference to `uint64_t` or `int64_t`. It could be even easier to pass `json &` and directly manipulate the value.

**Comment by nlohmann** [comment]
@J-e-r-e-m-i-a-h Do you need further assistance?

**Comment by J-e-r-e-m-i-a-h** [comment]
The example is wrong, by the way; the `square` function is improper (I made the post late an night). I just copied the parameters for a Windows function I'm using and then simplified it in order to post an example. Either way, I do require passing a variable by reference.

As of now, I'm trying to figure it out, I'll let you know if I can't figure out how to use get_ref :P I'm pretty sure your proposed solution is correct. Here's what I've got so far, let me know if I'm on the right track :)

```
if ( j[ "int" ].is_number_unsigned( ) )
    square( j[ "int" ].get_ref<int64_t&>( ) );
else if ( j[ "int" ].is_number_unsigned( ) )
    square( j[ "int" ].get_ref<uint64_t&>( ) );
```
